README.md
A flutter package which lets you use styled @mentions and #hashtags with twitter-like suggestions
in your TextField.

This was very much inspired by [flutter_parsed_text](https://pub.dev/packages/flutter_parsed_text) and
[flutter_mentions](https://pub.dev/packages/flutter_mentions), so credit to them!

# Table of contents

- [Installing](#installing)
- [Usage](#usage)
- [Custom Suggestion List](#custom-suggestion-list)

# Installing

### 1. Depend on it

Add this to your package's `pubspec.yaml` file:

```yaml
dependencies:
  flutter_parsed_text_field: ^0.1.0
```

### 2. Install it

You can install packages from the command line:

with `pub`:

```
$ pub get
```

with `Flutter`:

```
$ flutter pub get
```

### 3. Import it

Now in your `Dart` code, you can use:

```dart
import 'package:flutter_parsed_text_field/flutter_parsed_text_field.dart';
```

# Usage

### `FlutterParsedTextField` is a _Stateful Widget_ that extends TextField.

```dart
FlutterParsedTextField(
    controller: controller.flutterParsedTextEditingController,
    suggestionMatches: suggestionMatches,
    disableSuggestionOverlay: disableSuggestionOverlay,
    suggestionLimit: suggestionLimit,
    suggestionPosition: suggestionPosition,
    matchers: [],
)
```

Configurable properties:

- `controller` – extension of `TextEditingController` which styles your parsed text field
- `disableSuggestionOverlay` –  `true` if you don't want the built-in suggestion list to appear; `false` otherwise.
- `suggestionLimit` – number of suggestions to return
- `suggestionPosition` – should the suggestion list appear above or below the text field
- `matchers` - a list of `Matcher` which defines the triggers and suggestions

Callbacks:

- `suggestionMatches` – returns a list of the suggestions that are matched

### `Matcher`

```dart
Matcher<String>(
    trigger: "#",
    suggestions: ['BattleOfNewYork', 'InfinityGauntlet'],
    idProp: (hashtag) => hashtag,
    displayProp: (hashtag) => hashtag,
    style: const TextStyle(color: Colors.blue),
    stringify: (trigger, hashtag) => hashtag,
    alwaysHighlight: true,
    parse: (hashtagString) => hashtagString.substring(1),
)
```

Configurable properties:

- `trigger` - single character to trigger suggestions
- `suggestions` - list of suggestions
- `idProp` - get the id of the suggestion
- `displayProp` - get the display of the suggestion
- `style` - TextStyle for matches
- `stringify` - convert the suggestion into a parsable string
- `alwaysHighlight` - always apply style even if there isn't a matching suggestion
- `parse` - convert the parsable string into a suggestion

# Custom Suggestion List
If you dont want the built-in popup to appear, and instead display the suggestions elsewhere, you can!
Check /example/lib/custom-suggestion-list.dart

## Contributing
Pull requests are welcome. Please open an issue first to discuss what you would like to change.

## License
[MIT](https://choosealicense.com/licenses/mit/)

CHANGELOG.md
## v0.1.10 (May 18, 2022)
* upgrade deps

## v0.1.8 (Jan 18, 2022)
* only show overlay when textfield is focused

## v0.1.7 (Nov 24, 2021)
* add mention text to resultSort and finalResultSort

## v0.1.6 (Nov 18, 2021)
* enable matchers to sort results before and after limiting suggestions

## v0.1.5 (Nov 17, 2021)
* fix overlay throwing error when context is destroyed

## v0.1.4 (Nov 6, 2021)
* fix parsing not returning the trigger with always highlight
* change match order of always highlight

## v0.1.3 (Nov 03, 2021)
* enable parsing a stringified text (eg [@Ironman:3000]) into it's display form (eg @Ironman)

## v0.1.2 (Oct 13, 2021)
* position the built in suggestion popup to only take up the space it needs
* built in suggestion popup won't overflow the screen's bounds

## v0.1.1 (Oct 12, 2021)
* fix issue of stringifying when there are no suggestions

## v0.1.0 (Oct 12, 2021)
* initial release to allow multiple triggers and parsable text in a TextField 


lib/src/matcher.model.dart
part of flutter_parsed_text_field;

enum MatcherSearchStyle {
  /// does suggestion start with search criteria, case-sensitive
  startsWith,
  /// does suggestion contain the search criteria, case-sensitive
  contains,
  /// does suggestion start with search criteria, case-insensitive
  iStartsWith,
  /// does suggestion contain the search criteria, case-isensitive
  iContains,
}

class Matcher<T> {
  /// The single character that will cause the [suggestions] to start appearing
  final String trigger;

  /// A list of objects that should be used to populate the suggestions popup
  final List<T> suggestions;

  /// A function to return the id of the suggestion.
  ///
  /// Eg {userId: 'uid3000', displayName: 'Ironman'}
  /// idProp = (suggestion) => suggestion.userId
  final String Function(dynamic suggestion) idProp;

  /// A function to return the display of the suggestion.
  ///
  /// Eg {userId: 'uid3000', displayName: 'Ironman'}
  /// idProp = (suggestion) => suggestion.displayName
  final String Function(dynamic suggestion) displayProp;

  /// How the search logic should be applied
  final MatcherSearchStyle searchStyle;

  /// A function to return the prop for sorting the final results BEFORE [FlutterParsedTextField.suggestionLimit] is applied
  ///
  /// [text] is the text that this matchers is currently matching
  final int Function(String text, dynamic a, dynamic b)? resultSort;

  /// A function to return the prop for sorting the final results AFTER [FlutterParsedTextField.suggestionLimit] is applied
  ///
  /// [text] is the text that this matchers is currently matching
  final int Function(String text, dynamic a, dynamic b)? finalResultSort;

  /// A function which will stringify the suggestion
  ///
  /// Eg {userId: 'uid3000', displayName: 'Ironman'}
  /// stringify = (trigger, suggestion) => '[$trigger${suggestion.displayName}${suggestion.userId}]'
  final String Function(String trigger, dynamic suggestion) stringify;

  /// The [RegExp] that will match stringified suggestions and enable parsing
  ///
  /// Eg "Hey [[@Ironman:uid3000]]" => "Hey @Ironman"
  final RegExp parseRegExp;

  /// A function to parse the stringified suggestion back into an object.
  final T Function(RegExp regex, String stringifiedSuggestion) parse;

  /// When a suggestion has been tapped and added to the text field, it will be returned.
  /// This allows you to, say, update the #hashtag list when an @mention has been added.
  final Function(String trigger, dynamic suggestion)? onSuggestionAdded;

  /// The [TextStyle] to be applied to matches within the text field
  final TextStyle? style;

  /// Returns a widget that will be used in the suggestion popup's [ListView]
  final Widget Function(Matcher matcher, dynamic suggestion)? suggestionBuilder;

  /// True if the [style] should always be applied regardless if it matches a suggestion; false otherwise.
  ///
  /// This is good to set as true for #hashtags when there typically isn't a predefined list.
  /// This is good to set as false for @mentions when there typically is a predefined list.
  final bool alwaysHighlight;

  /// Starting index in the text field of the current match
  ///
  /// Eg "Hey @Iro" would be 4.
  int? indexOfMatch;

  /// The length of the partial match (including trigger)
  ///
  /// Eg "Hey @Ir" would be 3.
  int? lengthOfMatch;

  /// This is the core of what makes flutter_parsed_text_field work. This is where you define what [trigger]s suggestions,
  /// what those [suggestions] are, and how they are used.
  Matcher({
    required this.trigger,
    required this.suggestions,
    required this.idProp,
    required this.displayProp,
    this.searchStyle = MatcherSearchStyle.iContains,
    this.resultSort,
    this.finalResultSort,
    required this.stringify,
    required this.parse,
    required this.parseRegExp,
    this.onSuggestionAdded,
    this.style,
    this.suggestionBuilder,
    this.alwaysHighlight = false,
  }) : assert(trigger.length == 1);

  Type typeOf() => T;

  String get regexPattern {
    final regexes = [
      if (alwaysHighlight) '[A-Za-z0-9]+',
      ...suggestions.map(displayProp).map((s) => RegExp.escape(s)).sortedDescending(),
    ].map((s) => '$trigger$s');

    return regexes.isEmpty ? '' : '(${regexes.join('|')})';
  }
}


lib/src/flutter_parsed_text_field.dart
part of flutter_parsed_text_field;

enum SuggestionPosition { above, below }

class FlutterParsedTextField extends StatefulWidget {
  /// Extension of [TextEditingController] which applies your styles to the parsed text.
  /// This also allows you to get the stringified value of your text.
  ///
  /// Eg "Hey @Ironman" => "Hey [[@Ironman:uid3000]]"
  final FlutterParsedTextFieldController? controller;

  /// A list of [Matcher]s that the widget will use to parse your text.
  final List<Matcher> matchers;

  /// Returns a list of the suggestions that match your current [Matcher].
  ///
  /// Eg "@I" would return "@Ironman" and "@BlackWidow"
  final Function(Matcher? matcher, List matches)? suggestionMatches;

  /// True if you don't want the built-in suggestion list to appear; false otherwise.
  final bool disableSuggestionOverlay;

  /// The max number of suggestions returned when matching
  final int? suggestionLimit;

  /// The position of the built-in suggestion popup; above or below the text field
  final SuggestionPosition? suggestionPosition;

  /****************************************************************
      FLUTTER TEXT FIELD PROPS
   ***************************************************************/

  /// {@macro flutter.widgets.editableText.focusNode}
  final FocusNode? focusNode;

  /// {@macro flutter.widgets.editableText.decoration}
  final InputDecoration? decoration;

  /// {@macro flutter.widgets.editableText.keyboardType}
  final TextInputType? keyboardType;

  /// {@macro flutter.widgets.editableText.textInputAction}
  final TextInputAction? textInputAction;

  /// {@macro flutter.widgets.editableText.textCapitalization}
  final TextCapitalization textCapitalization;

  /// {@macro flutter.widgets.editableText.style}
  final TextStyle? style;

  /// {@macro flutter.widgets.editableText.strutStyle}
  final StrutStyle? strutStyle;

  /// {@macro flutter.widgets.editableText.textAlign}
  final TextAlign textAlign;

  /// {@macro flutter.widgets.editableText.textAlignVertical}
  final TextAlignVertical? textAlignVertical;

  /// {@macro flutter.widgets.editableText.textDirection}
  final TextDirection? textDirection;

  /// {@macro flutter.widgets.editableText.autofocus}
  final bool autofocus;

  /// {@macro flutter.widgets.editableText.obscuringCharacter}
  final String obscuringCharacter;

  /// {@macro flutter.widgets.editableText.obscureText}
  final bool obscureText;

  /// {@macro flutter.widgets.editableText.autocorrect}
  final bool autocorrect;

  /// {@macro flutter.widgets.editableText.smartDashesType}
  final SmartDashesType? smartDashesType;

  /// {@macro flutter.widgets.editableText.smartQuotesType}
  final SmartQuotesType? smartQuotesType;

  /// {@macro flutter.widgets.editableText.enableSuggestions}
  final bool enableSuggestions;

  /// {@macro flutter.widgets.editableText.maxLines}
  final int? maxLines;

  /// {@macro flutter.widgets.editableText.minLines}
  final int? minLines;

  /// {@macro flutter.widgets.editableText.expands}
  final bool expands;

  /// {@macro flutter.widgets.editableText.readOnly}
  final bool readOnly;

  /// {@macro flutter.widgets.editableText.toolbarOptions}
  final ToolbarOptions? toolbarOptions;

  /// {@macro flutter.widgets.editableText.showCursor}
  final bool? showCursor;

  /// {@macro flutter.widgets.editableText.maxLength}
  final int? maxLength;

  /// {@macro flutter.widgets.editableText.maxLengthEnforcement}
  final MaxLengthEnforcement? maxLengthEnforcement;

  /// {@macro flutter.widgets.editableText.onChanged}
  final ValueChanged<String>? onChanged;

  /// {@macro flutter.widgets.editableText.onEditingComplete}
  final VoidCallback? onEditingComplete;

  /// {@macro flutter.widgets.editableText.onSubmitted}
  final ValueChanged<String>? onSubmitted;

  /// {@macro flutter.widgets.editableText.onAppPrivateCommand}
  final AppPrivateCommandCallback? onAppPrivateCommand;

  /// {@macro flutter.widgets.editableText.inputFormatters}
  final List<TextInputFormatter>? inputFormatters;

  /// {@macro flutter.widgets.editableText.enabled}
  final bool? enabled;

  /// {@macro flutter.widgets.editableText.cursorWidth}
  final double cursorWidth;

  /// {@macro flutter.widgets.editableText.cursorHeight}
  final double? cursorHeight;

  /// {@macro flutter.widgets.editableText.cursorRadius}
  final Radius? cursorRadius;

  /// {@macro flutter.widgets.editableText.cursorColor}
  final Color? cursorColor;

  /// {@macro flutter.widgets.editableText.selectionHeightStyle}
  final ui.BoxHeightStyle selectionHeightStyle;

  /// {@macro flutter.widgets.editableText.selectionWidthStyle}
  final ui.BoxWidthStyle selectionWidthStyle;

  /// {@macro flutter.widgets.editableText.keyboardAppearance}
  final Brightness? keyboardAppearance;

  /// {@macro flutter.widgets.editableText.scrollPadding}
  final EdgeInsets scrollPadding;

  /// {@macro flutter.widgets.editableText.enableInteractiveSelection}
  final bool enableInteractiveSelection;

  /// {@macro flutter.widgets.editableText.selectionControls}
  final TextSelectionControls? selectionControls;

  /// {@macro flutter.widgets.editableText.dragStartBehavior}
  final DragStartBehavior dragStartBehavior;

  /// {@macro flutter.widgets.editableText.onTap}
  final GestureTapCallback? onTap;

  /// {@macro flutter.widgets.editableText.mouseCursor}
  final MouseCursor? mouseCursor;

  /// {@macro flutter.widgets.editableText.buildCounter}
  final InputCounterWidgetBuilder? buildCounter;

  /// {@macro flutter.widgets.editableText.scrollPhysics}
  final ScrollPhysics? scrollPhysics;

  /// {@macro flutter.widgets.editableText.scrollController}
  final ScrollController? scrollController;

  /// {@macro flutter.widgets.editableText.autofillHints}
  final Iterable<String>? autofillHints;

  /// {@macro flutter.widgets.editableText.restorationId}
  final String? restorationId;

  /// {@macro flutter.widgets.editableText.enableIMEPersonalizedLearning}
  final bool enableIMEPersonalizedLearning;

  const FlutterParsedTextField({
    Key? key,
    this.controller,
    required this.matchers,
    this.suggestionMatches,
    this.disableSuggestionOverlay = false,
    this.suggestionLimit,
    this.suggestionPosition = SuggestionPosition.below,
    // flutter text field props
    this.focusNode,
    this.decoration = const InputDecoration(),
    this.keyboardType,
    this.textInputAction,
    this.textCapitalization = TextCapitalization.none,
    this.style,
    this.strutStyle,
    this.textAlign = TextAlign.start,
    this.textAlignVertical,
    this.textDirection,
    this.readOnly = false,
    this.toolbarOptions,
    this.showCursor,
    this.autofocus = false,
    this.obscuringCharacter = '•',
    this.obscureText = false,
    this.autocorrect = true,
    this.smartDashesType,
    this.smartQuotesType,
    this.enableSuggestions = true,
    this.maxLines = 1,
    this.minLines,
    this.expands = false,
    this.maxLength,
    this.maxLengthEnforcement,
    this.onChanged,
    this.onEditingComplete,
    this.onSubmitted,
    this.onAppPrivateCommand,
    this.inputFormatters,
    this.enabled,
    this.cursorWidth = 2.0,
    this.cursorHeight,
    this.cursorRadius,
    this.cursorColor,
    this.selectionHeightStyle = ui.BoxHeightStyle.tight,
    this.selectionWidthStyle = ui.BoxWidthStyle.tight,
    this.keyboardAppearance,
    this.scrollPadding = const EdgeInsets.all(20.0),
    this.dragStartBehavior = DragStartBehavior.start,
    this.enableInteractiveSelection = true,
    this.selectionControls,
    this.onTap,
    this.mouseCursor,
    this.buildCounter,
    this.scrollController,
    this.scrollPhysics,
    this.autofillHints,
    this.restorationId,
    this.enableIMEPersonalizedLearning = true,
  }) : super(key: key);

  @override
  FlutterParsedTextFieldState createState() => FlutterParsedTextFieldState();
}

class FlutterParsedTextFieldState extends State<FlutterParsedTextField> {
  late FlutterParsedTextFieldController _controller;

  final LayerLink _layerLink = LayerLink();
  OverlayEntry? _suggestionOverlay;

  /// Apply the selected [suggestion] from the [matcher] to the text field.
  /// This will replace the partial text with the suggestion.
  ///
  /// Eg text field of "Hey @Ir", then applying will result in "Hey @Ironman"
  void applySuggestion<T>({
    required Matcher matcher,
    required T suggestion,
  }) {
    assert(matcher.indexOfMatch != null && matcher.lengthOfMatch != null);

    final replacement = '${matcher.trigger}${matcher.displayProp(suggestion)} ';
    _controller.text = _controller.value.text.replaceRange(
      matcher.indexOfMatch!,
      matcher.indexOfMatch! + matcher.lengthOfMatch!,
      replacement,
    );

    _controller.selection = TextSelection.fromPosition(TextPosition(offset: matcher.indexOfMatch! + replacement.length));

    if (matcher.onSuggestionAdded != null) {
      matcher.onSuggestionAdded!(matcher.trigger, suggestion);
    }
  }

  void _showSuggestionsOverlay<T>({
    required Matcher matcher,
    required List<T> suggestions,
  }) {
    final renderBox = context.findRenderObject() as RenderBox?;

    if (renderBox != null) {
      _hideSuggestionOverlay();

      _suggestionOverlay = OverlayEntry(
        builder: (context) {
          if (!renderBox.attached) {
            return const SizedBox();
          }

          var globalOffset = renderBox.localToGlobal(Offset.zero);
          var size = renderBox.size;
          var spaceBelow = MediaQuery.of(context).size.height - MediaQuery.of(context).viewInsets.bottom - globalOffset.dy - size.height - 16;
          var spaceAbove = globalOffset.dy - MediaQuery.of(context).padding.top - kToolbarHeight - 16;

          return Positioned(
            width: size.width,
            height: widget.suggestionPosition == SuggestionPosition.above ? spaceAbove : spaceBelow,
            child: CompositedTransformFollower(
              link: _layerLink,
              followerAnchor: widget.suggestionPosition == SuggestionPosition.above ? Alignment.bottomCenter : Alignment.topCenter,
              targetAnchor: widget.suggestionPosition == SuggestionPosition.above ? Alignment.topCenter : Alignment.bottomCenter,
              showWhenUnlinked: false,
              offset: Offset(0, widget.suggestionPosition == SuggestionPosition.above ? -8 : 8),
              child: Align(
                alignment: widget.suggestionPosition == SuggestionPosition.above ? Alignment.bottomCenter : Alignment.topCenter,
                child: Material(
                  elevation: 4,
                  child: ListView.builder(
                    padding: EdgeInsets.zero,
                    shrinkWrap: true,
                    itemCount: suggestions.length,
                    itemBuilder: (BuildContext context, int index) {
                      final suggestion = suggestions[index];

                      if (matcher.suggestionBuilder != null) {
                        return matcher.suggestionBuilder!(
                          matcher,
                          suggestion,
                        );
                      }

                      return ListTile(
                        title: Text('${matcher.trigger}${matcher.displayProp(suggestion)}'),
                        onTap: () => applySuggestion(
                          matcher: matcher,
                          suggestion: suggestion,
                        ),
                      );
                    },
                  ),
                ),
              ),
            ),
          );
        },
      );

      Overlay.of(context)!.insert(_suggestionOverlay!);
    }
  }

  void _hideSuggestionOverlay() {
    _suggestionOverlay?.remove();
    _suggestionOverlay = null;
  }

  void _suggestionListener() {
    final cursorPos = _controller.selection.baseOffset;

    if (cursorPos > 0) {
      final text = _controller.value.text;
      final lastIndexOfSpace = text.lastIndexOf(RegExp(r'\s'), cursorPos - 1);

      var indexOfMatch = lastIndexOfSpace + 1;

      final token = text.substring(indexOfMatch).split(RegExp(r'(\s)')).first;
      var lengthOfMatch = token.length;

      if (token.isNotEmpty) {
        final matchers = widget.matchers.where((matcher) => matcher.trigger == token[0]);

        if (matchers.length > 1) {
          throw 'Multiple matchers match $token';
        }

        if (matchers.length == 1) {
          final search = token.substring(1);
          final matcher = matchers.first;

          var matchedSuggestions = matcher.suggestions.where((e) {
            switch (matcher.searchStyle) {
              case MatcherSearchStyle.startsWith:
                return matcher.displayProp(e).startsWith(search);
              case MatcherSearchStyle.contains:
                return matcher.displayProp(e).contains(search);
              case MatcherSearchStyle.iStartsWith:
                return matcher.displayProp(e).toLowerCase().startsWith(search.toLowerCase());
              case MatcherSearchStyle.iContains:
                return matcher.displayProp(e).toLowerCase().contains(search.toLowerCase());
            }
          }).toList();

          if (matcher.resultSort != null) {
            matchedSuggestions.sort((a, b) => matcher.resultSort!(search, a, b));
          }

          if (widget.suggestionLimit != null) {
            matchedSuggestions = matchedSuggestions.take(widget.suggestionLimit!).toList();
          }

          if (matcher.finalResultSort != null) {
            matchedSuggestions.sort((a, b) => matcher.finalResultSort!(search, a, b));
          }

          matcher.indexOfMatch = indexOfMatch;
          matcher.lengthOfMatch = lengthOfMatch;

          if (widget.suggestionMatches != null) {
            widget.suggestionMatches!(matcher, matchedSuggestions);
          }

          if (!widget.disableSuggestionOverlay) {
            _showSuggestionsOverlay(
              matcher: matcher,
              suggestions: matchedSuggestions,
            );
          }

          return;
        }
      }
    }

    if (widget.suggestionMatches != null) {
      widget.suggestionMatches!(null, []);
    }

    _hideSuggestionOverlay();
  }

  void _applySuggestionListener() {
    applySuggestion(
      matcher: context.read<SuggestionApplier>().matcher!,
      suggestion: context.read<SuggestionApplier>().suggestionToApply!,
    );

    _hideSuggestionOverlay();
  }

  @override
  void initState() {
    super.initState();

    _controller = widget.controller ?? FlutterParsedTextFieldController();
    _controller.matchers = widget.matchers;

    _controller.addListener(_suggestionListener);
  }

  @override
  void didUpdateWidget(covariant FlutterParsedTextField oldWidget) {
    _controller.matchers = widget.matchers;

    super.didUpdateWidget(oldWidget);
  }

  @override
  void didChangeDependencies() {
    super.didChangeDependencies();

    context.watch<SuggestionApplier?>()?.removeListener(_applySuggestionListener);
    context.watch<SuggestionApplier?>()?.addListener(_applySuggestionListener);
  }

  @override
  Widget build(BuildContext context) {
    return CompositedTransformTarget(
      link: _layerLink,
      child: Focus(
        focusNode: widget.focusNode,
        onFocusChange: (hasFocus) {
          if (!hasFocus) {
            _hideSuggestionOverlay();
          }
        },
        child: TextField(
          controller: widget.controller,
          decoration: widget.decoration,
          keyboardType: widget.keyboardType,
          textInputAction: widget.textInputAction,
          textCapitalization: widget.textCapitalization,
          style: widget.style,
          strutStyle: widget.strutStyle,
          textAlign: widget.textAlign,
          textAlignVertical: widget.textAlignVertical,
          textDirection: widget.textDirection,
          readOnly: widget.readOnly,
          toolbarOptions: widget.toolbarOptions,
          showCursor: widget.showCursor,
          autofocus: widget.autofocus,
          obscuringCharacter: widget.obscuringCharacter,
          obscureText: widget.obscureText,
          autocorrect: widget.autocorrect,
          smartDashesType: widget.smartDashesType,
          smartQuotesType: widget.smartQuotesType,
          enableSuggestions: widget.enableSuggestions,
          maxLines: widget.maxLines,
          minLines: widget.minLines,
          expands: widget.expands,
          maxLength: widget.maxLength,
          maxLengthEnforcement: widget.maxLengthEnforcement,
          onChanged: widget.onChanged,
          onEditingComplete: widget.onEditingComplete,
          onSubmitted: widget.onSubmitted,
          onAppPrivateCommand: widget.onAppPrivateCommand,
          inputFormatters: widget.inputFormatters,
          enabled: widget.enabled,
          cursorWidth: widget.cursorWidth,
          cursorHeight: widget.cursorHeight,
          cursorRadius: widget.cursorRadius,
          cursorColor: widget.cursorColor,
          selectionHeightStyle: widget.selectionHeightStyle,
          selectionWidthStyle: widget.selectionWidthStyle,
          keyboardAppearance: widget.keyboardAppearance,
          scrollPadding: widget.scrollPadding,
          dragStartBehavior: widget.dragStartBehavior,
          enableInteractiveSelection: widget.enableInteractiveSelection,
          selectionControls: widget.selectionControls,
          onTap: widget.onTap,
          mouseCursor: widget.mouseCursor,
          buildCounter: widget.buildCounter,
          scrollController: widget.scrollController,
          scrollPhysics: widget.scrollPhysics,
          autofillHints: widget.autofillHints,
          restorationId: widget.restorationId,
          enableIMEPersonalizedLearning: widget.enableIMEPersonalizedLearning,
        ),
      ),
    );
  }

  @override
  void dispose() {
    _controller.removeListener(_suggestionListener);

    super.dispose();
  }
}


example/lib/custom_suggestion_list.dart
import 'package:flutter/material.dart';
import 'package:flutter_parsed_text_field/flutter_parsed_text_field.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Parsed Text Field',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(),
    );
  }
}

class Avenger {
  final String userId;
  final String displayName;

  Avenger({
    required this.userId,
    required this.displayName,
  });
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({
    Key? key,
  }) : super(key: key);

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final flutterParsedTextFieldController = FlutterParsedTextFieldController();

  /// this is what sends the selected suggestion down to FlutterParsedTextField
  final suggestionApplier = SuggestionApplier();

  Matcher? matcher;
  List? matchedSuggestions;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Flutter Parsed Text Field'),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            ChangeNotifierProvider<SuggestionApplier>.value(
              value: suggestionApplier,
              builder: (_, __) => FlutterParsedTextField(
                controller: flutterParsedTextFieldController,
                disableSuggestionOverlay: true,
                suggestionLimit: 5,
                suggestionMatches: (matcher, suggestions) {
                  /// returns the matcher and suggestions based on your input
                  setState(() {
                    this.matcher = matcher;
                    matchedSuggestions = suggestions;
                  });
                },
                matchers: [
                  Matcher<Avenger>(
                    trigger: "@",
                    suggestions: [
                      Avenger(userId: '3000', displayName: 'Ironman'),
                      Avenger(userId: '4000', displayName: 'Hulk'),
                      Avenger(userId: '5000', displayName: 'Black Widow'),
                    ],
                    idProp: (avenger) => avenger.userId,
                    displayProp: (avenger) => avenger.displayName,
                    style: const TextStyle(color: Colors.red),
                    stringify: (trigger, avenger) {
                      return '[$trigger${avenger.displayName}:${avenger.userId}]';
                    },
                    parseRegExp: RegExp(r"\[(@([^\]]+)):([^\]]+)\]"),
                    parse: (regex, avengerString) {
                      final avenger = regex.firstMatch(avengerString);

                      if (avenger != null) {
                        return Avenger(
                          userId: avenger.group(3)!,
                          displayName: avenger.group(2)!,
                        );
                      }

                      throw 'Avenger not found';
                    },
                  ),
                ],
              ),
            ),
            if (matcher != null && matchedSuggestions != null)
            /// ugly UI, but it illustrates the point
              Expanded(
                child: Padding(
                  padding: const EdgeInsets.all(8.0),
                  child: Material(
                    elevation: 4,
                    child: ListView.builder(
                      itemCount: matchedSuggestions!.length,
                      shrinkWrap: true,
                      itemBuilder: (BuildContext context, int index) {
                        final suggestion = matchedSuggestions![index] as Avenger;

                        return ListTile(
                          title: Text('@${suggestion.displayName}'),
                          subtitle: Text(suggestion.userId),
                          onTap: () {
                            /// send the tapped suggestion down to FlutterParsedTextField for processing
                            suggestionApplier.updateSuggestion(
                              matcher: matcher!,
                              suggestionToApply: suggestion,
                            );
                          },
                        );
                      },
                    ),
                  ),
                ),
              ),
            TextButton(
              child: const Text('Clear'),
              onPressed: () => flutterParsedTextFieldController.clear(),
            )
          ],
        ),
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_parsed_text_field/flutter_parsed_text_field.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Parsed Text Field',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(),
    );
  }
}

class Avenger {
  final String userId;
  final String displayName;

  Avenger({
    required this.userId,
    required this.displayName,
  });
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({
    Key? key,
  }) : super(key: key);

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final flutterParsedTextFieldController = FlutterParsedTextFieldController();

  String addedAvenger = '';

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Flutter Parsed Text Field'),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            TextButton(
              onPressed: () => FocusScope.of(context).unfocus(),
              child: const Text('Unfocus'),
            ),
            Text('Recently Added Avenger: $addedAvenger'),
            FlutterParsedTextField(
              controller: flutterParsedTextFieldController,
              suggestionMatches: (matcher, suggestions) {},
              disableSuggestionOverlay: false,
              suggestionLimit: 5,
              matchers: [
                Matcher<Avenger>(
                  trigger: "@",
                  suggestions: [
                    Avenger(userId: '3000', displayName: 'Ironman'),
                    Avenger(userId: '4000', displayName: 'Hulk'),
                    Avenger(userId: '5000', displayName: 'Black Widow'),
                  ],
                  idProp: (avenger) => avenger.userId,
                  displayProp: (avenger) => avenger.displayName,
                  style: const TextStyle(color: Colors.red),
                  stringify: (trigger, avenger) {
                    return '[$trigger${avenger.displayName}:${avenger.userId}]';
                  },
                  parseRegExp: RegExp(r"\[(@([^\]]+)):([^\]]+)\]"),
                  parse: (regexp, avengerString) {
                    final avenger = regexp.firstMatch(avengerString);

                    if (avenger != null) {
                      return Avenger(
                        userId: avenger.group(3)!,
                        displayName: avenger.group(2)!,
                      );
                    }

                    throw 'Avenger not found';
                  },
                  onSuggestionAdded: (trigger, avenger) {
                    setState(() => addedAvenger = avenger.displayName);
                  },
                ),
                Matcher<String>(
                  trigger: "#",
                  suggestions: ['BattleOfNewYork', 'InfinityGauntlet'],
                  idProp: (hashtag) => hashtag,
                  displayProp: (hashtag) => hashtag,
                  style: const TextStyle(color: Colors.blue),
                  stringify: (trigger, hashtag) => hashtag,
                  alwaysHighlight: true,
                  parseRegExp: RegExp(r'(#([\w]+))'),
                  parse: (regex, hashtagString) => hashtagString,
                ),
              ],
            ),
            TextButton(
              child: const Text('Clear'),
              onPressed: () => flutterParsedTextFieldController.clear(),
            )
          ],
        ),
      ),
    );
  }
}
